docs/guide/custom-login-requirements.md

```diff
@@ -32,15 +32,112 @@ While [most password complexity rules are often counterproductive](https://arste
+<AposCodeBlock>
+  ```javascript
+  module.exports = {
+    requirements(self) {
+      return {
+        add: {
+          RequirementName: {
+            phase: 'afterSubmit',
+            async props(req) {
+              return {
+                propName: propValue
+              };
+            },
+            async verify(req) {
+              if (!verifyThisCodeIsCorrect(req.body.requirement.RequirementName)) {
+                throw self.apos.error('invalid', 'appropriate message');
+              }
+            }
+          }
+        }
+      };
+    }
+  };
+  ```
+  <template v-slot:caption>
+    modules/@apostrophecms/login/index.js
+  </template>
+</AposCodeBlock>
+
+## The browser side
+
+Each distinct requirement also has browser-side code, which is implemented as a Vue component. As explained in the [custom UI guide](../custom-ui.md), Vue components intended for the admin UI (including login requirements) must be placed in the `ui/apos/components` subdirectory of a module in the project. For this purpose they are typically placed in the `modules/@apostrophecms/login/ui/apos/components` module at project level, or in an npm module that enhances `@apostrophecms/login` via `improve`.
+
+The developer is responsible for the appearance of the component. For `beforeSubmit` requirements, the component will appear at the bottom of the login form itself. For `afterSubmit` requirements, it will appear on its own, "fading up" after the Login button is clicked, although password verification has not happened yet. For `afterPasswordVerified` requirements, the appearance is the same as for `afterSubmit`, although password verification has been completed at this point. If there are multiple `afterSubmit` and `afterPasswordVerified` requirements, the user will see them presented one at a time, with the `afterSubmit` requirements completed first.
+
+Each component is responsible for:
+
+* Presenting its own UI, if any. Apostrophe will handle "fading up" `afterSubmit` and `afterPasswordVerified` components and displaying any suitable outer framing.
+* Accepting the properties of the object returned by the server-side `props(req)` function as props.
+* Emitting a `done` event with a payload providing proof that the requirement has been met. This proof is accessible to the `verify(req)` function on the server side as `req.body.requirements.RequirementName`. In the case of `afterSubmit` and `afterPasswordVerified` requirements, the `done` event should not be emitted until the user has indicated their response is complete in some way.
+
+While there is no fixed structure for the Vue components, a typical outline looks like:
+
+<AposCodeBlock>
+  ```javascript
+  <template>
+    <fieldset>
+      <label>What code did you receive?</label>
+      <p>Hint: {{ hint }}</p>
+      <input type="text" v-model="value" />
+      <button type="button" @click="go">OK</button>
+    </fieldset>
+  </template>
+  <script>
+  export default {
+    emits: [ 'done' ],
```

Comment:
Shouldn't we include the block event as well?

Reply:
We would, but it doesn't yet exist. It was a bit hypothetical and there's no proven need yet.